python/imzml_reader.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+
+import numpy as np
+import pandas as pd
+from pyimzml.ImzMLParser import ImzMLParser
+import zarr
+
+import argparse
+
+
+class IMSDataset:
+    """Converts IMS data stored as imzML into columnar or ndarray formats
+
+    :param imzml_file: path to `.imzML` file.
+    :param ibd_file: path to associated `.ibd` file.
+    :param micro_res: microscopy resolution in nm (used for scaling).
+    :param ims_res: IMS resolution in nm (used for scaling).
+    """
+
+    def __init__(self, imzml_file, ibd_file, micro_res, ims_res):
+        # When passing the ibd path explicitly,
+        # the file object must be opened manually
+        self.parser = ImzMLParser(
+            filename=imzml_file, ibd_file=open(ibd_file, "rb")
+        )
+        self.micro_res = micro_res
+        self.ims_res = ims_res
+        self.ims_px_in_micro = ims_res / micro_res  # scaling factor
+
+        mz_lengths = self.parser.mzLengths
+        if not (mz_lengths.count(mz_lengths[0]) == len(mz_lengths)):
+            raise ValueError(
+                "The number of m/z is not the same at each coordinate."
+            )
+        self.mzs, _ = self.parser.getspectrum(0)
+
+    def __get_min_max_coords(self):
+        coords = np.array(self.parser.coordinates)
+        x_min, y_min, _ = np.min(coords, axis=0)
+        x_max, y_max, _ = np.max(coords, axis=0)
+        return x_min, y_min, x_max, y_max
+
+    def __format_mzs(self, precision=4):
+        return np.round(self.mzs, precision).astype(str)
+
+    def to_columnar(self, dtype="uint32"):
+        coords = np.array(self.parser.coordinates)
+        x, y, _ = coords.T
+
+        coords_df = pd.DataFrame(
+            {
+                "x": x,
+                "y": y,
+                "micro_x_topleft": x * self.ims_px_in_micro
+                - self.ims_px_in_micro,
+                "micro_y_topleft": y * self.ims_px_in_micro
+                - self.ims_px_in_micro,
+                "micro_px_width": np.repeat(self.ims_px_in_micro, len(coords)),
+            },
+            dtype=dtype,
+        )
+
+        intensities = np.zeros((len(coords_df), len(self.mzs)))
+        for i in range(len(coords)):
+            _, coord_intensities = self.parser.getspectrum(i)
+            intensities[i, :] = coord_intensities
```

Comment:
Would something like this work?
```python
intensities = numpy.asarray(
  [self.parser.getspectrum(i)[1] for i in range(len(coords))]
)
```

Reply:
I think that would work, yes, but I decided to pre-allocate the large numpy array and then fill it rather than creating a list of lists and coercing that into a numpy array. I believe this offers one contiguous piece of memory and will be more performant to write to (maybe I'm wrong here). For toy datasets it doesn't really matter in terms of performance, but the number of rows in the dataframe gets pretty large (especially when we start to deal with 3D).

Anecdotally, I'd previously used a generator to create the dataframe and saw a speed up from ~10sec to ~3sec when pre-allocating the memory for the numpy array.

Reply:
I can run some tests if you'd like.

Reply:
cool: retesting isn't necessary: That seems plausible. Can you add a note explaining that pre-allocation is for performance? Thanks!

Reply:
Will do. One thing I think is worth mentioning here,`vitessce-data` does not actually call this method currently. I ported it from the previous repository since this is the processed format the Spraggin's lab provides on Globus. Do you think we should have a fixture for this as well? Or should I remove this from `vitessce-data`?

Reply:
Your call: If it's likely to get used, eventually, let's leave it in?